internal/stores/hostdb.go

```diff
@@ -181,3 +195,236 @@ func NewJSONHostDB(dir string) (*JSONHostDB, modules.ConsensusChangeID, error) {
+
+// NewSQLHostDB uses a given Dialector to connect to a SQL database.  NOTE: Only
+// pass migrate=true for the first instance of SQLHostDB if you connect via the
+// same Dialector multiple times.
+func NewSQLHostDB(conn gorm.Dialector, migrate bool) (*SQLHostDB, modules.ConsensusChangeID, error) {
+	db, err := gorm.Open(conn, &gorm.Config{})
+	if err != nil {
+		return nil, modules.ConsensusChangeID{}, err
+	}
+
+	if migrate {
+		// Create the tables.
+		if err := db.AutoMigrate(&host{}); err != nil {
+			return nil, modules.ConsensusChangeID{}, err
+		}
+		if err := db.AutoMigrate(&interaction{}); err != nil {
+			return nil, modules.ConsensusChangeID{}, err
+		}
+		if err := db.AutoMigrate(&announcement{}); err != nil {
+			return nil, modules.ConsensusChangeID{}, err
+		}
+		if err := db.AutoMigrate(&consensusInfo{}); err != nil {
+			return nil, modules.ConsensusChangeID{}, err
+		}
+	}
+
+	// Get latest consensus change ID or init db.
+	var ci consensusInfo
+	err = db.Where(&consensusInfo{ID: consensusInfoID}).
+		Attrs(consensusInfo{
+			ID:   consensusInfoID,
+			CCID: modules.ConsensusChangeBeginning[:],
+		}).
+		FirstOrCreate(&ci).Error
+	if err != nil {
+		return nil, modules.ConsensusChangeID{}, err
+	}
+	var ccid modules.ConsensusChangeID
+	copy(ccid[:], ci.CCID)
+
+	return &SQLHostDB{
+		staticDB: db,
+	}, ccid, nil
+}
+
+// Host returns information about a host.
+func (db *SQLHostDB) Host(hostKey consensus.PublicKey) (hostdb.Host, error) {
+	var h host
+	tx := db.staticDB.Where(&host{PublicKey: hostKey[:]}).
+		Preload("Interactions").
+		Preload("Announcements").
+		Find(&h)
+	if tx.Error == nil && tx.RowsAffected == 0 {
+		return hostdb.Host{}, ErrHostNotFound
+	}
+	return h.Host(), tx.Error
+}
+
+// Hosts returns up to max hosts that have not been interacted with since
+// the specified time.
+func (db *SQLHostDB) Hosts(notSince time.Time, max int) ([]hostdb.Host, error) {
+	lastInteraction := func(h hostdb.Host) time.Time {
+		if len(h.Interactions) == 0 {
+			return time.Time{}
+		}
+		return h.Interactions[len(h.Interactions)-1].Timestamp
+	}
+	hosts, err := db.SelectHosts(math.MaxInt, func(h hostdb.Host) bool { return true })
+	if err != nil {
+		return nil, err
+	}
+	for _, host := range hosts {
+		if len(hosts) == max {
+			break
+		} else if lastInteraction(host).Before(notSince) {
+			hosts = append(hosts, host)
+		}
+	}
+	return hosts, nil
+}
+
+// RecordInteraction records an interaction with a host. If the host is not in
+// the store, a new entry is created for it.
+func (db *SQLHostDB) RecordInteraction(hostKey consensus.PublicKey, hi hostdb.Interaction) error {
+	return db.staticDB.Transaction(func(tx *gorm.DB) error {
+		// Create a host if it doesn't exist yet.
+		if err := tx.FirstOrCreate(&host{}, &host{PublicKey: hostKey[:]}).Error; err != nil {
+			return err
+		}
+
+		// Create an interaction.
+		return tx.Create(&interaction{
+			Host:      hostKey[:],
+			Timestamp: hi.Timestamp.UTC(), // use UTC in DB
```

Comment:
generally considered good practice to store all timestamps as UTC in a database 🤷‍♂️

Reply:
Ok I did some testing and I need to keep the `UTC` in the test code (due to the reflect.DeepEqual) but can strip it from the database code. It will be turned into UTC implicitly.

Reply:
isn't it better to be explicit?

Reply:
did you find out how the timestamp is encoded? Does it include the timezone or not?

Reply:
I think it's fine to not be explicit here since gorm should handle all of the conversion between native types and database types. It's only an issue in testing because it's not safe to use `reflect.DeepEqual` on `time.Time` in this way. A regular `time.Equals` should work and all the other time related methods as well.

Reply:
ok. Given that the encoding contains timezone info, I'm onboard with converting values to UTC before storing them in the DB.